core/src/main/scala/org/apache/spark/resource/ResourceUtils.scala

```diff
@@ -0,0 +1,191 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.resource
+
+import java.io.File
+import java.nio.file.{Files, Paths}
+
+import scala.util.control.NonFatal
+
+import org.json4s.DefaultFormats
+import org.json4s.jackson.JsonMethods._
+
+import org.apache.spark.{SparkConf, SparkException}
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config._
+import org.apache.spark.util.Utils.executeAndGetOutput
+
+/**
+ * Resource identifier.
+ * @param componentName spark.driver / spark.executor / spark.task
+ * @param resourceName  gpu, fpga, etc
+ */
+private[spark] case class ResourceID(componentName: String, resourceName: String) {
+  def confPrefix: String = s"$componentName.resource.$resourceName." // with ending dot
+  def amountConf: String = s"$confPrefix${ResourceUtils.AMOUNT}"
+  def discoveryScriptConf: String = s"$confPrefix${ResourceUtils.DISCOVERY_SCRIPT}"
+  def vendorConf: String = s"$confPrefix${ResourceUtils.VENDOR}"
+}
+
+private[spark] case class ResourceRequest(
+    id: ResourceID,
+    amount: Int,
+    discoveryScript: Option[String],
+    vendor: Option[String])
+
+private[spark] case class TaskResourceRequirement(resourceName: String, amount: Int)
+
+/**
+ * Case class representing allocated resource addresses for a specific resource.
+ * Cluster manager uses the JSON serialization of this case class to pass allocated resource info to
+ * driver and executors. See the ``--resourcesFile`` option there.
+ */
+private[spark] case class ResourceAllocation(id: ResourceID, addresses: Seq[String]) {
+  def toResourceInformation: ResourceInformation = {
+    new ResourceInformation(id.resourceName, addresses.toArray)
+  }
+}
+
+private[spark] object ResourceUtils extends Logging {
+
+  // config suffixes
+  val DISCOVERY_SCRIPT = "discoveryScript"
+  val VENDOR = "vendor"
+  // user facing configs use .amount to allow to extend in the future,
+  // internally we currently only support addresses, so its just an integer count
+  val AMOUNT = "amount"
+
+  def parseResourceRequest(sparkConf: SparkConf, resourceId: ResourceID): ResourceRequest = {
+    val settings = sparkConf.getAllWithPrefix(resourceId.confPrefix).toMap
+    val amount = settings.getOrElse(AMOUNT,
+      throw new SparkException(s"You must specify an amount for ${resourceId.resourceName}")
+    ).toInt
+    val discoveryScript = settings.get(DISCOVERY_SCRIPT)
+    val vendor = settings.get(VENDOR)
+    ResourceRequest(resourceId, amount, discoveryScript, vendor)
+  }
+
+  def listResourceIds(sparkConf: SparkConf, componentName: String): Seq[ResourceID] = {
+    sparkConf.getAllWithPrefix(s"$componentName.resource.").map { case (key, _) =>
+      key.substring(0, key.indexOf('.'))
+    }.toSet.toSeq.map(name => ResourceID(componentName, name))
+  }
+
+  def parseAllResourceRequests(
+      sparkConf: SparkConf,
+      componentName: String): Seq[ResourceRequest] = {
+    listResourceIds(sparkConf, componentName).map { id =>
```

Comment:
This would call `sparkConf.getAllWithPrefix` multiple times, we may reduce the call to only once and manually turn the `Map[String, String]` to `Seq[ResourceRequest]`. I understand this is tradeoff between code readability and performance, I'm just worried the config number can be big thus we want to reduce the call of `getAllWithPrefix`. I'd prefer to leave a TODO here and consider the improvement in the future.

Reply:
This is a one-time only call for each service. I think we should optimize SparkConf and getAllWithPrefix instead.